webclient/webclient/src/main/java/io/helidon/webclient/SingleWithContext.java

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.webclient;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Flow;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+
+import io.helidon.common.context.Context;
+import io.helidon.common.context.Contexts;
+import io.helidon.common.reactive.CompletionAwaitable;
+import io.helidon.common.reactive.Multi;
+import io.helidon.common.reactive.Single;
+
+class SingleWithContext<T> implements Single<T> {
```

Comment:
Shorter version could look like this:
```java
    public <T> Single<T> singleWithContext(Single<T> single, Context context) {
        return Single.create(subscriber -> single.subscribe(new Flow.Subscriber<T>() {
            @Override
            public void onSubscribe(Flow.Subscription subscription) {
                Contexts.runInContext(context, () -> subscriber.onSubscribe(subscription));
            }

            @Override
            public void onNext(T item) {
                Contexts.runInContext(context, () -> subscriber.onNext(item));
            }

            @Override
            public void onError(Throwable throwable) {
                Contexts.runInContext(context, () -> subscriber.onError(throwable));
            }

            @Override
            public void onComplete() {
                Contexts.runInContext(context, subscriber::onComplete);
            }
        }));
    }
```

Reply:
Could be tweaked for propagating also to the upstream over the backpressure, but that could lead to unexpected/unwanted effects on different kinds of publishers